packages/@sanity/form-builder/src/inputs/Slug/SlugInput.tsx

```diff
@@ -42,169 +28,119 @@ type Props = {
-        const {loading, inputText} = this.state
-        const hasSourceField = type.options && type.options.source
-        const formFieldProps = {
-          label: type.title,
-          description: type.description,
-          level: level,
-          markers,
-          presence,
-          labelFor: this._inputId,
-        }
-        const errors = markers.filter(isValidationErrorMarker)
-        return (
-          <ChangeIndicatorCompareValueProvider
-            value={value?.current}
-            compareValue={compareValue?.current}
-          >
-            <FormField {...formFieldProps}>
-              <div className={styles.wrapper}>
-                <div className={styles.input}>
-                  <TextInput
-                    inputId={this._inputId}
-                    ref={this.setTextInput}
-                    customValidity={errors.length > 0 ? errors[0].item.message : ''}
-                    disabled={loading}
-                    onChange={this.handleChange}
-                    value={typeof inputText === 'string' ? inputText : value.current}
-                    readOnly={readOnly}
-                  />
-                </div>
-                {hasSourceField && (
-                  <Button
-                    className={styles.button}
-                    disabled={readOnly || loading || !this.hasSource()}
-                    loading={loading}
-                    onClick={this.handleGenerateSlug}
-                    kind="simple"
-                  >
-                    Generate
-                  </Button>
-                )}
-              </div>
-            </FormField>
-          </ChangeIndicatorCompareValueProvider>
-        )
-      }
+  const [generateState, handleGenerateSlug] = useAsyncCallback(() => {
+    if (!sourceField) {
+      return Promise.reject(
+        new Error(`Source is missing. Check source on type "${type.name}" in schema`)
+      )
     }
+
+    return getNewFromSource(sourceField, getValuePath(), document)
+      .then((newFromSource) => slugify(newFromSource || '', type))
+      .then((newSlug) => updateSlug(newSlug))
+  }, [getValuePath, updateSlug, document, type])
+
+  const isUpdating = generateState?.status === 'pending'
+
+  return (
+    <ChangeIndicatorCompareValueProvider
+      value={value?.current}
+      compareValue={compareValue?.current}
+    >
+      <FormField
+        title={type.title}
+        description={type.description}
+        level={level}
+        __unstable_markers={markers}
+        __unstable_presence={presence}
+        inputId={inputId}
+      >
+        <Stack space={3}>
+          <Flex>
+            <Box flex={1}>
+              <TextInput
+                id={inputId}
+                ref={forwardedRef}
+                customValidity={errors.length > 0 ? errors[0].item.message : ''}
+                disabled={isUpdating}
+                onChange={React.useCallback((event) => updateSlug(event.currentTarget.value), [
```

Comment:
Assigning a variable for this (e.g. `handleChange`) would de-nest and make this a bit easier to read.